packages/block-editor/src/hooks/use-bindings-attributes.js

```diff
@@ -22,87 +25,255 @@ import { unlock } from '../lock-unlock';
-										placeholder;
-								}
-							}
-
-							if ( metaValue ) {
-								updatedAttributes[ attributeName ] = metaValue;
-							}
-						}
-					}
-				);
+/**
+ * Based on the given block name,
+ * check if it is possible to bind the block.
+ *
+ * @param {string} blockName - The block name.
+ * @return {boolean} Whether it is possible to bind the block to sources.
+ */
+export function canBindBlock( blockName ) {
+	return blockName in BLOCK_BINDINGS_ALLOWED_BLOCKS;
+}
+
+/**
+ * Based on the given block name and attribute name,
+ * check if it is possible to bind the block attribute.
+ *
+ * @param {string} blockName     - The block name.
+ * @param {string} attributeName - The attribute name.
+ * @return {boolean} Whether it is possible to bind the block attribute.
+ */
+export function canBindAttribute( blockName, attributeName ) {
+	return (
+		canBindBlock( blockName ) &&
+		BLOCK_BINDINGS_ALLOWED_BLOCKS[ blockName ].includes( attributeName )
+	);
+}
+
+/**
+ * This component is responsible for detecting and
+ * propagating data changes from the source to the block.
+ *
+ * @param {Object}   props                   - The component props.
+ * @param {string}   props.attrName          - The attribute name.
+ * @param {any}      props.attrValue         - The attribute value.
+ * @param {string}   props.blockName         - The block name.
+ * @param {Object}   props.blockProps        - The block props with bound attribute.
+ * @param {Object}   props.source            - Source handler.
+ * @param {Object}   props.args              - The arguments to pass to the source.
+ * @param {Function} props.onPropValueChange - The function to call when the attribute value changes.
+ * @return {null}                              Data-handling component. Render nothing.
+ */
+const BindingConnector = ( {
+	args,
+	attrName,
+	attrValue,
+	blockName,
+	blockProps,
+	source,
+	onPropValueChange,
+} ) => {
+	const { placeholder, value: propValue } = source.useSource(
+		blockProps,
+		args
+	);
+
+	const updateBoundAttibute = useCallback(
+		( newAttrValue, prevAttrValue ) => {
+			/*
+			 * If the attribute is a RichTextData instance,
+			 * (core/paragraph, core/heading, core/button, etc.)
+			 * compare its HTML representation with the new value.
+			 *
+			 * To do: it looks like a workaround.
+			 * Consider improving the attribute and metadata fields types.
+			 */
+			if ( prevAttrValue instanceof RichTextData ) {
+				// Bail early if the Rich Text value is the same.
+				if ( prevAttrValue.toHTMLString() === newAttrValue ) {
+					return;
+				}
+
+				/*
+				 * To preserve the value type,
+				 * convert the new value to a RichTextData instance.
+				 */
+				newAttrValue = RichTextData.fromHTMLString( newAttrValue );
+			}
+
+			if ( prevAttrValue === newAttrValue ) {
+				return;
 			}
 
-			return (
+			onPropValueChange?.( { [ attrName ]: newAttrValue } );
+		},
+		[ attrName, onPropValueChange ]
+	);
+
+	useEffect( () => {
```

Comment:
The fact that this uses an effect means that there's a small time frame where the value passed to the block is incorrect. I think ultimately that could be a problem. So in other words, if there's a way to achieve the current PR without effects (useMemo and useCallback are fine), that would be preferred.

Reply:
Makes sense, I've started to address this by removing the `useEffect` and calling the `updateBoundAttibute()` directly in the render function. 

I'm seeing an [infinite loop](https://react.dev/errors/185?invariant=185) so likely I'm missing a check somewhere to prevent the state update in some circumstances. cc @SantosGuillamot as this is probably the same issue as what you've seen.

Reply:
calling updateBoundAttribute won't work either, we can't update local state in render functions, they are supposed to be immutable. I think the ideal would be to find a way to have this:

```js 
const { attributes, setAttributes } = useBoundAttributes( originalAttributes, originalSetAttributes )

return <BlockEditor attributes={ attributes } setAttributes={ setAttributes } />
```

Reply:
I'm not sure I have a good solution about how to implement this, it might mean chaining the `useSource` API. That said, I'd be ok with the "useEffect" for initial version but we should keep an eye on it and not open the API publicly until it's proven bug free.

Reply:
>The fact that this uses an effect means that there's a small time frame where the value passed to the block is incorrect. I think ultimately that could be a problem.

Indeed, I believe the `useEffect` is causing this "blinking" effect in the query loop: [link](https://github.com/WordPress/gutenberg/pull/58895#issuecomment-1959802730).

> cc @SantosGuillamot as this is probably the same issue as what you've seen.

Yes, not using `useEffect`, with the current approach, creates an infinite loop.

Reply:
> Indeed, I believe the useEffect is causing this "bhttps://github.com/WordPress/gutenberg/pull/58895#issuecomment-1959802730ing" effect in the query loop: link.

Let's explore enhancements for it separately, given that it would be great to fix the blinking in WP 6.5. At the same time, it doesn't seem as big issue, as the fact that `useSource` doesn't behave like a React hook causing hard to track issues.

Reply:
It should be possible to avoid using `useEffect()` and call the setter in the render function. Basically using [this pattern](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes). But we should do it carefully because with more complex logic it's easy to introduce infinite loops ♾️ .

Reply:
I opened a draft PR following this approach in https://github.com/WordPress/gutenberg/pull/59443